modem: modem_cellular: Configurable MTU for CMUX

SeppoTakalo opened the pull request:
Allow configuring MTU for CMUX.
Some AT manual and specification define this as a
frame size. Linux ldattach default to 127 bytes,
3GPP TS 27.010 defaults to 31.

We should limit our CMUX frames to a size that
remote end is capable of handling.
Linux silently drops oversized frames.

Also, deprecate MODEM_CELLULAR_CMUX_MAX_FRAME_SIZE as this was only limiting a buffer sizes, and resulted CMUX frames to be capped to same value.
Use MODEM_CMUX_WORK_BUFFER_SIZE and MODEM_CMUX_MTU instead.

For reference:
Linux kernel `n_gsm.c` driver on receiving side: https://elixir.bootlin.com/linux/v6.13.6/source/drivers/tty/n_gsm.c#L2885
You can see that all frames larger than MTU/MRU are silently dropped.

`ldattach(8)` utility from Util-linux package: https://github.com/util-linux/util-linux/blob/master/sys-utils/ldattach.c#L260-L276 you can see that it sets default to 127 and there is no command line parameters to configure this.

SeppoTakalo committed:
modem: modem_cellular: Configurable MTU for CMUX

Allow configuring MTU for CMUX.
Some AT manual and specification define this as a
frame size. Linux ldattach default to 127 bytes,
3GPP TS 27.010 defaults to 31.

We should limit our CMUX frames to a size that
remote end is capable of handling.
Linux silently drops oversized frames.

Also, remove MODEM_CELLULAR_CMUX_MAX_FRAME_SIZE as
this was only limiting a buffer sizes, and resulted
CMUX frames to be capped to same value.
Use MODEM_CMUX_WORK_BUFFER_SIZE and MODEM_CMUX_MTU instead.

Also rename CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES to
CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZE as it is now
only used as a Chat module. DLCI pipes use
CONFIG_MODEM_CMUX_WORK_BUFFER_SIZE.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

jukkar commented on drivers/modem/Kconfig.cellular:
```diff
@@ -43,11 +43,10 @@ config MODEM_CELLULAR_UART_BUFFER_SIZES
-	default 32 if DT_HAS_U_BLOX_LARA_R6_ENABLED
-	default 128
+	int "[Deprecated] The maximum frame size in bytes."
+	default 0
```
You should probably add here
```
select DEPRECATED
```
I think one does not have to write `[Deprecated]` to prompt as selecting the DEPRECATED will give user enough info.

SeppoTakalo replied:
int values cannot select another value.

jukkar replied:
:facepalm: Indeed, never mind then

bjarki-andreasen reviewed (dismissed):
I like it :)

SeppoTakalo commented:
Updated the PR to contain `CONFIG_MODEM_CMUX_MTU=64` for CMUX tests as these already used patterns bigger than 31.

Also added a unittest for splitting transmitted data to size of MTU.

tomi-font commented on drivers/modem/Kconfig.cellular:
```diff
@@ -43,11 +43,10 @@ config MODEM_CELLULAR_UART_BUFFER_SIZES
-	default 32 if DT_HAS_U_BLOX_LARA_R6_ENABLED
-	default 128
+	int "[Deprecated] The maximum frame size in bytes."
+	default 0
```
Why `default 0`? Why have a default value at all anymore?

tomi-font commented on drivers/modem/Kconfig.cellular:
```diff
@@ -43,11 +43,10 @@ config MODEM_CELLULAR_UART_BUFFER_SIZES
 config MODEM_CELLULAR_CMUX_MAX_FRAME_SIZE
-	int "The maximum CMUX frame size in bytes."
-	default 32 if DT_HAS_U_BLOX_LARA_R6_ENABLED
-	default 128
+	int "[Deprecated] The maximum frame size in bytes."
```
You deprecate the Kconfig option but in practice it doesn't do anything anymore, so I wouldn't call that deprecation. At that stage why not remove it directly? A user would be misguided thinking this Kconfig option can still be set while in practice it will have no effect.

SeppoTakalo replied:
Fair point.. Am I allowed to just remove this?

I might break existing builds... but also, if someone misses deprecation note from changelog, their app is unusable anyway.

I will remove the Kconfig.

tomi-font commented on doc/releases/release-notes-4.2.rst:
Migration guide entry rather?

tomi-font commented on subsys/modem/Kconfig:
```diff
@@ -31,8 +31,19 @@ if MODEM_CMUX
+	default 31
+	help
+	  Maximum Transmission Unit (MTU) size for the CMUX module.
+	  Linux ldattach default to 127 bytes, 3GPP TS 27.010 defaults to 31.
```
```suggestion
	  Linux ldattach defaults to 127 bytes, 3GPP TS 27.010 to 31.
```

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -90,16 +90,16 @@ struct modem_cellular_data {
 	/* Points to dlci2_pipe or NULL. Used for shutdown script if not NULL */
 	struct modem_pipe *cmd_pipe;
-	uint8_t dlci1_receive_buf[CONFIG_MODEM_CELLULAR_CMUX_MAX_FRAME_SIZE];
+	uint8_t dlci1_receive_buf[CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES];
```
`CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES` was meant for buffers used with the chat module, so for AT commands.
The driver turns DLCI1 into the PPP channel when everything's up and running, and frames may be more than 1000 bytes in size depending on the PPP MTU.
This wouldn't play out well in that scenario. I think this needs to use a different Kconfig option.

SeppoTakalo replied:
At least for nRF91 this driver executed `AT#XCMUX=1` command, which turn DLCI1 as AT channel and channel 2 as PPP.

I'm a bit unsure where the channel selection really is, as all the drivers are generated by macros.

tomi-font replied:
The `AT#XCMUX=1` command is for turning on CMUX. Then at a later point (as part of the dial chat script) the `AT#XCMUX=2` command is issued, which will result in DLCI1 being the PPP channel.

We can say that the channel selection happens here in the driver (`modem_ppp_attach()`): 
https://github.com/zephyrproject-rtos/zephyr/blob/1708574d0a78e687ff832f7c165de3bdaff54cae/drivers/modem/modem_cellular.c#L1026-L1034

SeppoTakalo replied:
You are right. I updated all dlci buffers to use `CONFIG_MODEM_CMUX_WORK_BUFFER_SIZE` and it seem to be now working OK.

SeppoTakalo committed:
modem: cmux: Do not return error on low buffer space

If we end up writing zero bytes to cmux output, we can return
zero instead of -ENOMEM as it would break various modules when
using small buffers. For example modem_chat.c does not tolerate
-ENOMEM but handles zero OK.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo commented:
* Removed the Kconfig entirely and added a migration note instead.
* Use `CONFIG_MODEM_CMUX_WORK_BUFFER_SIZE` as this is the size of basic unit all CMUX layers work on. PPP and Chat have their own internal buffers defined elsewhere.
* Minor fix to modem_cmux.c to return zero instead of `-ENOMEM` when the receiving buffer is too small. This was causing `modem_chat.c` to abort.

jukkar commented on subsys/modem/Kconfig:
```diff
@@ -31,8 +31,19 @@ if MODEM_CMUX
+	default 38
+	help
+	  Size of the work buffer used by the CMUX module.
+	  Recommended size is MODEM_CMUX_MTU + 7 (CMUX header size).
```
Should there be a build assert that verifies that this size is larger than the cmux mtu?

SeppoTakalo replied:
I did not test this, but my understanding is that both CMUX and PPP are able to work even if the input buffer size is only one byte. It would be just so unoptimal that data drops would break the packets.

Therefore I would not like to add asserts, until I really know that lower values break things.

tomi-font commented on samples/net/cellular_modem/boards/nrf9160dk_nrf52840.conf:
I think that all the `CONFIG_MODEM_CELLULAR_CMUX_MAX_FRAME_SIZE` that were explicitly set need to be replaced by appropriate `CONFIG_MODEM_CMUX_WORK_BUFFER_SIZE` and `CONFIG_MODEM_CMUX_MTU` values?
Their defaults are rather low and I imagine this will cause issues if the modem sends CMUX frames that are bigger than the CMUX MTU on the driver side (e.g. in the case of SLM with PPP).

SeppoTakalo replied:
I went through AT manuals for all modems that I found and compared against AT commands we send on `modem_cellular.c`. None matched what was set on `CONFIG_MODEM_CELLULAR_CMUX_MAX_FRAME_SIZE` so I assume existing configuration was not based on any real test scenario.

tomi-font commented on drivers/modem/Kconfig.cellular:
```diff
@@ -42,13 +42,6 @@ config MODEM_CELLULAR_UART_BUFFER_SIZES
-	help
-	  This value affects the size of buffers used to receive and transmit CMUX frames.
-
 config MODEM_CELLULAR_CHAT_BUFFER_SIZES
 	int "The size of the buffers used for the chat scripts in bytes."
 	default 128
```
Then this is only used for one buffer anymore, so it's a bit strange to have it use the plural.
`dlci2_receive_buf` could maybe be left to keep using this as it's only used for chat scripts? It would also reduce RAM usage in case of big CMUX MTU.

SeppoTakalo replied:
I don't touch the Chat module in this PR so I would like not to extend the scope of this.
Otherwise this fix would just keep expanding and we never get it in.

tomi-font reviewed (changes requested):
The commit message of `modem: modem_cellular: Configurable MTU for CMUX` should be updated to reflect the last changes.

tomi-font reviewed (commented):
> However, based on my current understanding is that none of the cases that previously set CONFIG_MODEM_CELLULAR_CMUX_MAX_FRAME_SIZE would have actually worked. Some set the size to 1500, but on AT command, set MTU to 127. This probably allowed ping to work (small packet) but all else failed.

1500 was only used for SLM which has a big MTU by default so no problem with that one. For the two other boards it's possible that bigger frames could have caused issues, though PPP has its own MTU so CMUX frames might have been split based on that.

SeppoTakalo commented on subsys/modem/Kconfig:
```diff
@@ -29,10 +29,34 @@ config MODEM_CMUX
+		      DT_HAS_SQN_GM02S_ENABLED)
+	help
+	  Use the default MTU size of 127 bytes for the CMUX module on certain modems.
+	  This must match AT commands on modem_cellular.c
```
Thanks. Fixed.

SeppoTakalo commented:
> 1500 was only used for SLM which has a big MTU by default so no problem with that one. For the two other boards it's possible that bigger frames could have caused issues, though PPP has its own MTU so CMUX frames might have been split based on that.

Yes, SLM worked as both of the side were based on same CMUX implementation. But all other modems used CMUX frame size that is different than what we might have passed to them. And as we did not drop any frames in receive side, this would have caused very weird issues. So sending 1500 sized PPP would fail, but receiving anything works, as it was capped by modem to 127.

> I was just proposing to keep the size of dlci2_receive_buf in modem_cellular.c unchanged (CONFIG_MODEM_CELLULAR_CHAT_BUFFER_SIZES). Or then this Kconfig option could be renamed and its description updated to reflect the changes.

I see. I renamed the Kconfig as I want to keep all DLCI buffers to same as CMUX size. It seem to work OK and we then save lots of RAM.

```
# samples/net/cellular_modem on nrf9160dk/nrf52840
# This PR
Memory region         Used Size  Region Size  %age Used
           FLASH:      173316 B         1 MB     16.53%
             RAM:       59308 B       256 KB     22.62%
        IDT_LIST:          0 GB        32 KB      0.00%

# main branch
[308/308] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:      179348 B         1 MB     17.10%
             RAM:       65260 B       256 KB     24.89%
        IDT_LIST:          0 GB        32 KB      0.00%
```

SeppoTakalo commented:
Rebased on top of main and fixed the conflict on `migration-guide-4.2.rst`

SeppoTakalo commented:
Please merge this soon, as I would otherwise have to constantly rebase it on top of main, because changelog keeps changing and I get conflicts.

tomi-font commented:
> I see. I renamed the Kconfig as I want to keep all DLCI buffers to same as CMUX size. It seem to work OK and we then save lots of RAM.

Uhm, weird numbers considering your changes (especially regarding ROM), are you sure your branch is based off the same main as what you are using for comparison?

Also, of course you'll get big RAM savings for `nrf9160dk/nrf52840` by changing the buffer sizes from 1500 to 38, but I think this is gonna be very sub-ideal for the bandwidth as every PPP frame will be very small due to the CMUX MTU.
For the SLM-based board targets I'd actually support keeping a bigger CMUX MTU in the sample.

SeppoTakalo commented:
> Uhm, weird numbers considering your changes (regarding ROM), are you sure your branch is based off the same main as what you are using for comparison?

Not weird numbers. I took a look of why this happens.
`struct modem_cellular_data` contains all buffers so when all modem instances are defined using `static struct modem_cellular_data MODEM_CELLULAR_INST_NAME(data, inst) = {...}` then all fields of that struct are stored in flash, and then copied into `.data`section.

So even though we only initialize some of the fields, the whole structure is stored. So all buffers are going to be zero initialized.

SeppoTakalo commented:
> sub-ideal for the bandwidth as every PPP frame will be very small due to the CMUX MTU.

I'm unsure if I follow this correctly. The CMUX MTU does not affect PPP MTU.
One PPP frame might need multiple CMUX frames to be send.

For performance wise, all buffer sizes should be optimized only once you define your final baud rate. So it is pointless to try to tweak them here.

SeppoTakalo committed:
modem: modem_cellular: Migration guide for CMUX

Add migration guide for configuring MTU for CMUX.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo commented:
Sorry for changing this PR:
I separated the migration guide to its own commit so I'm able to pull these changes in nRF Connect SDK.

Otherwise I would end up conflicting the migration guide again when cherry-picking.

tomi-font commented:
> Not weird numbers. I took a look of why this happens.
struct modem_cellular_data contains all buffers so when all modem instances are defined using static struct modem_cellular_data MODEM_CELLULAR_INST_NAME(data, inst) = {...} then all fields of that struct are stored in flash, and then copied into .datasection.
>
> So even though we only initialize some of the fields, the whole structure is stored. So all buffers are going to be zero initialized.

Oh wow. That's bad. Wondering what could be done to avoid that. :thinking: 

> I'm unsure if I follow this correctly. The CMUX MTU does not affect PPP MTU.
One PPP frame might need multiple CMUX frames to be send.

Yeah sorry I was too simplistic in what I said. The PPP frames being sent across many small CMUX frames will reduce the MCU <=> modem bandwidth (and increase at least a little bit the processing overhead). But I don't disagree that this needs to be tuned according to every use case and configuration.

tomi-font commented:
> Sorry for changing this PR:
I separated the migration guide to its own commit so I'm able to pull these changes in nRF Connect SDK.
>
> Otherwise I would end up conflicting the migration guide again when cherry-picking.

The diff didn't change so the reviews weren't dismissed. :slightly_smiling_face:
